### ISSUE TITLE
discover: IndexError: list index out of range

### ISSUE THREAD

**Issue by ArturGajowy** [VirtusLab/git-machete#2]
Hi! Thx for making this very promising tool!

I hit the following error when doing `git machete discover -v`:

```
artur@y510p:~/projects/rchain$ git machete discover -v
git branch
git reflog --format=%H:%gs RHOL-537 --
git log --max-count=10 --format=%H 35dab3bebad9b73679497df9a1e3a2a1cf2bcc65
git reflog --format=%H:%gs RHOL-689-bigger-ints --
git rev-parse --abbrev-ref RHOL-689-bigger-ints@{upstream}
git reflog --format=%H:%gs add_coursier --
git rev-parse --abbrev-ref add_coursier@{upstream}
git reflog --format=%H:%gs artur/remove_unused_var --
git reflog --format=%H:%gs backtest --
git rev-parse --abbrev-ref backtest@{upstream}
git reflog --format=%H:%gs cleanups_on_the_go --
git reflog --format=%H:%gs dev --
git rev-parse --abbrev-ref dev@{upstream}
git reflog --format=%H:%gs dev-fix-erroneous-connective --
git rev-parse --abbrev-ref dev-fix-erroneous-connective@{upstream}
git reflog --format=%H:%gs dev-ground-connectives --
git rev-parse --abbrev-ref dev-ground-connectives@{upstream}
git reflog --format=%H:%gs huge_string_concat --
git rev-parse --abbrev-ref huge_string_concat@{upstream}
git reflog --format=%H:%gs master --
git reflog --format=%H:%gs maximum_bipartite_match --
git reflog --format=%H:%gs mbm_memo --
git rev-parse --abbrev-ref mbm_memo@{upstream}
git reflog --format=%H:%gs mbm_perf_fix --
git rev-parse --abbrev-ref mbm_perf_fix@{upstream}
git reflog --format=%H:%gs mbm_v2 --
git rev-parse --abbrev-ref mbm_v2@{upstream}
git reflog --format=%H:%gs refactors --
git rev-parse --abbrev-ref refactors@{upstream}
git reflog --format=%H:%gs reminder_syntax_maps_and_sets --
git rev-parse --abbrev-ref reminder_syntax_maps_and_sets@{upstream}
git reflog --format=%H:%gs rho_parse --
git rev-parse --abbrev-ref rho_parse@{upstream}
git reflog --format=%H:%gs scalawave --
git rev-parse --abbrev-ref scalawave@{upstream}
git reflog --format=%H:%gs stable_marriage --
git reflog --format=%H:%gs stable_marriage_no_method_sort --
git reflog --format=%H:%gs stable_marriage_rebase --
git log --format=%H 35dab3bebad9b73679497df9a1e3a2a1cf2bcc65
git log --max-count=10 --format=%H 2da3512897752a88028e0e317a9161a4f540cb43
git rev-parse --abbrev-ref RHOL-537@{upstream}
git log --format=%H 2da3512897752a88028e0e317a9161a4f540cb43
git log --max-count=10 --format=%H 8c165feeac22f09d56182014c5d5393006dabb1c
git log --format=%H 8c165feeac22f09d56182014c5d5393006dabb1c
Traceback (most recent call last):
  File "/usr/local/bin/git-machete", line 1325, in <module>
    discover_tree()
  File "/usr/local/bin/git-machete", line 518, in discover_tree
    u = infer_upstream(b, condition=lambda u: get_root_of(u) != b)
  File "/usr/local/bin/git-machete", line 482, in infer_upstream
    start_points = commits or [reflog(b)[-1][0]]
IndexError: list index out of range
```

My guess would be the `--max-count=10` being used?

**Comment by PawelLipski** [comment]
Hi @ArturGajowy :) thanks! 

Could you run the following against your repository and send the output:
`git for-each-ref --format='%(refname:short)' refs/heads | xargs -t -L1 git --no-pager reflog --format=%H`

My suspicion is that reflog of some of the branches (most likely `RHOL-689-bigger-ints`) is empty for whatever reason... most likely due to gc/reflog expiration. If that's actually the case, the fix should be pretty easy. I'd also add the checks for that circumstances across the entire code.

**Comment by ArturGajowy** [comment]
I've since done a very aggressive GC so were not able to reproduce with any meaningful data. Here's another set of test data though:


```
artur@y510p:~/projects/rchain$ git machete discover -v
git branch
git reflog --format=%H:%gs add_coursier --
git log --max-count=10 --format=%H 63a0785d1542a7f4419e7413c354df53ec4db06b
git reflog --format=%H:%gs dev --
git rev-parse --abbrev-ref dev@{upstream}
git reflog --format=%H:%gs huge_string_concat --
git reflog --format=%H:%gs master --
git reflog --format=%H:%gs mbm_memo --
git rev-parse --abbrev-ref mbm_memo@{upstream}
git reflog --format=%H:%gs package_object --
git rev-parse --abbrev-ref package_object@{upstream}
git reflog --format=%H:%gs post_memo_refactors --
git reflog --format=%H:%gs rho_parse --
git log --format=%H 63a0785d1542a7f4419e7413c354df53ec4db06b
git log --max-count=10 --format=%H 118c9b712cf02952955c375d86a5ec53424276f5
git log --format=%H 118c9b712cf02952955c375d86a5ec53424276f5
git log --max-count=10 --format=%H e33559543183910025da052dde60d8f86788d2a5
git log --format=%H e33559543183910025da052dde60d8f86788d2a5
Traceback (most recent call last):
  File "/usr/local/bin/git-machete", line 1325, in <module>
    discover_tree()
  File "/usr/local/bin/git-machete", line 518, in discover_tree
    u = infer_upstream(b, condition=lambda u: get_root_of(u) != b)
  File "/usr/local/bin/git-machete", line 482, in infer_upstream
    start_points = commits or [reflog(b)[-1][0]]
IndexError: list index out of range
```

```
artur@y510p:~/projects/rchain$ git for-each-ref --format='%(refname:short)' refs/heads | xargs -t -L1 git --no-pager reflog --format=%H
git --no-pager reflog --format=%H add_coursier 
b3c66f50f8ed73fff80453235a223e830076a243
63a0785d1542a7f4419e7413c354df53ec4db06b
git --no-pager reflog --format=%H dev 
f77881068e77b947e6be6b09accae6f5378a60a6
118c9b712cf02952955c375d86a5ec53424276f5
a2c4d8445b76449bfb29797a99b255c3bcd5e7d8
4d15bb72ea7df00b7e5c0dc7f48cea8b62a6e829
git --no-pager reflog --format=%H huge_string_concat 
532684d67dae54624c9647e9f12966bf0c31e390
e33559543183910025da052dde60d8f86788d2a5
git --no-pager reflog --format=%H master 
git --no-pager reflog --format=%H mbm_memo 
90a2f047922f6a34a195b78608c57ce78ca48c18
3803faa92a01342a6245c26ae261d0c4e8788020
7d250625a22c3a802c289783138b1113bde56666
git --no-pager reflog --format=%H package_object 
b37f11b6c9c627d998a40acc2ebcb50fc87bdb82
ff6b1e14af3521d7d975397536287986bfcb1b80
f5d10e80a0fbaf8b6e477075e60cdd0b1b2da5d8
git --no-pager reflog --format=%H post_memo_refactors 
1633c5a41dff6c279c0ee2a955066e1a20b9910f
6445dac7ef111ec4621c8a3a7c34257e56fcf689
9d8c4bf811a558f89f41c2fe5eedf4f94528330f
9d8c4bf811a558f89f41c2fe5eedf4f94528330f
git --no-pager reflog --format=%H rho_parse 
e4a0f9298ba266fbea94ddb5330ea56de801d787
1144203eb9f4b8f0555c4828b99a0585d0193357
```

**Comment by ArturGajowy** [comment]
I've checked out your last commit, but it doesn't seem to fix the issue:

```
artur@y510p:~/projects/rchain$ cat `which git-machete` | grep expi
    # Reflog of any branch can be empty due to e.g. reflog expiration.
artur@y510p:~/projects/rchain$ git machete discover -v
git branch
git reflog --format=%H:%gs add_coursier --
git log --max-count=10 --format=%H 63a0785d1542a7f4419e7413c354df53ec4db06b
git reflog --format=%H:%gs dev --
git rev-parse --abbrev-ref dev@{upstream}
git reflog --format=%H:%gs huge_string_concat --
git reflog --format=%H:%gs master --
git reflog --format=%H:%gs mbm_memo --
git rev-parse --abbrev-ref mbm_memo@{upstream}
git reflog --format=%H:%gs package_object --
git rev-parse --abbrev-ref package_object@{upstream}
git reflog --format=%H:%gs post_memo_refactors --
git reflog --format=%H:%gs rho_parse --
git log --format=%H 63a0785d1542a7f4419e7413c354df53ec4db06b
git log --max-count=10 --format=%H 118c9b712cf02952955c375d86a5ec53424276f5
git log --format=%H 118c9b712cf02952955c375d86a5ec53424276f5
git log --max-count=10 --format=%H e33559543183910025da052dde60d8f86788d2a5
git log --format=%H e33559543183910025da052dde60d8f86788d2a5
Traceback (most recent call last):
  File "/usr/local/bin/git-machete", line 1326, in <module>
    discover_tree()
  File "/usr/local/bin/git-machete", line 519, in discover_tree
    u = infer_upstream(b, condition=lambda u: get_root_of(u) != b)
  File "/usr/local/bin/git-machete", line 483, in infer_upstream
    start_points = commits or (reflog and [reflog(b)[-1][0]])
IndexError: list index out of range
```

Otherwise, I'd like you to know configuring the .git/machete and using update/slide-out/traverse has been a blast for me :D That's the tool I've dreamt of ever since I've worked at prestodb/presto :joy:

**Comment by PawelLipski** [comment]
Awww thanks! <3
 
I experimentally simplified the upstream inference logic for corner cases like yours, pls pull the most recent `bugfix/empty-reflog`.

**Comment by ArturGajowy** [comment]
This seems to fix it for me :)

```
artur@y510p:~/projects/rchain$ git machete discover 
Discovered tree of branch dependencies:

  dev
  │ 
  └─add_coursier
  │ 
  └─fix_warnings
  │ 
  └─hydra_gitignore
  │ 
  └─mbm_memo
  │ │ 
  │ └─huge_string_concat
  │ │ 
  │ └─post_memo_refactors
  │   │ 
  │   └─stacksafe_sorter
  │ 
  └─package_object
  │ 
  └─rho_parse

  master (behind origin)

Save the above tree to '.git/machete'? The existing definition file will be backed up as '.git/machete~' (y[es]/e[dit]/n[o]) n

#replied "no"

artur@y510p:~/projects/rchain$ git branch
  add_coursier
  dev
  fix_warnings
* huge_string_concat
  hydra_gitignore
  master
  mbm_memo
  package_object
  post_memo_refactors
  rho_parse
  stacksafe_sorter

artur@y510p:~/projects/rchain$ git machete status
  dev
  │ 
  └─mbm_memo
  │ │ 
  │ └─post_memo_refactors
  │   │ 
  │   └─stacksafe_sorter
  │     │ 
  │     └─huge_string_concat
  │ 
  └─rho_parse

artur@y510p:~/projects/rchain$ git machete list unmanaged 
add_coursier
fix_warnings
hydra_gitignore
master
package_object
artur@y510p:~/projects/rchain$ 
```

**Comment by ArturGajowy** [comment]
Maybe with the exception, that the discovered position of huge_string_concat is wrong - but it doesn't crash, which is better :)

**Comment by PawelLipski** [comment]
Fixed as of [v2.5.3](https://github.com/VirtusLab/git-machete/releases/tag/v2.5.3).

**Comment by PawelLipski** [comment]
Created a separate issue (#4) for your latest comment regarding incorrectly inferred parent branch.
